lib/common/options.c

```diff
@@ -38,6 +38,445 @@ pcmk__cli_help(char cmd)
+            "rules. However, it will also recheck the cluster after this "
+            "amount of inactivity, to evaluate rules with date specifications "
+            "and serve as a fail-safe for certain types of scheduler bugs."),
+    },
+    {
+        PCMK_OPT_FENCE_REACTION, NULL, "select", "stop, panic",
+        "stop", NULL,
+        N_("How a cluster node should react if notified of its own fencing"),
+        N_("A cluster node may receive notification of its own fencing if "
+            "fencing is misconfigured, or if fabric fencing is in use that "
+            "doesn't cut cluster communication. Use \"stop\" to attempt to "
+            "immediately stop Pacemaker and stay stopped, or \"panic\" to "
+            "attempt to immediately reboot the local node, falling back to "
+            "stop on failure."),
+    },
+    {
+        PCMK_OPT_ELECTION_TIMEOUT, NULL, "time", NULL,
+        "2min", pcmk__valid_interval_spec,
+        N_("*** Advanced Use Only ***"),
+        N_("Declare an election failed if it is not decided within this much "
+            "time. If you need to adjust this value, it probably indicates "
+            "the presence of a bug."),
+    },
+    {
+        PCMK_OPT_SHUTDOWN_ESCALATION, NULL, "time", NULL,
+        "20min", pcmk__valid_interval_spec,
+        N_("*** Advanced Use Only ***"),
+        N_("Exit immediately if shutdown does not complete within this much "
+            "time. If you need to adjust this value, it probably indicates "
+            "the presence of a bug."),
+    },
+    {
+        PCMK_OPT_JOIN_INTEGRATION_TIMEOUT, "crmd-integration-timeout", "time",
+            NULL,
+        "3min", pcmk__valid_interval_spec,
+        N_("*** Advanced Use Only ***"),
+        N_("If you need to adjust this value, it probably indicates "
+            "the presence of a bug."),
+    },
+    {
+        PCMK_OPT_JOIN_FINALIZATION_TIMEOUT, "crmd-finalization-timeout",
+            "time", NULL,
+        "30min", pcmk__valid_interval_spec,
+        N_("*** Advanced Use Only ***"),
+        N_("If you need to adjust this value, it probably indicates "
+            "the presence of a bug."),
+    },
+    {
+        PCMK_OPT_TRANSITION_DELAY, "crmd-transition-delay", "time", NULL,
+        "0s", pcmk__valid_interval_spec,
+        N_("*** Advanced Use Only *** "
+            "Enabling this option will slow down cluster recovery under all "
+            "conditions"),
+        N_("Delay cluster recovery for this much time to allow for additional "
+            "events to occur. Useful if your configuration is sensitive to "
+            "the order in which ping updates arrive."),
+    },
+    {
+        /* @TODO Actually must be a timeout (parsable by crm_get_msec()); can't
+         * be an ISO 8601 duration, which pcmk__valid_interval_spec() accepts.
+         */
+        PCMK_OPT_STONITH_WATCHDOG_TIMEOUT, NULL, "time", NULL,
+        "0", pcmk__valid_interval_spec,
```

Comment:
Actually `pcmk__get_sbd_timeout()` parses it with `crm_get_msec()`. Interval specs are a superset, so I eventually want to get around to upgrading parsed timeouts to interval specs, but this one will be problematic because cluster and remote nodes pass it back and forth.

Reply:
That's my point in the TODO comment. `pcmk__valid_sbd_timeout()` parses it with `crm_get_msec()`, but the validation function (`pcmk__valid_interval_spec()`) accepts a superset of the allowed values.

Or are you just suggesting I update the comment to point to `pcmk__get_sbd_timeout()` instead of `pcmk__valid_sbd_timeout()` for clarity?

Reply:
Let's create a `pcmk__valid_timeout()` for `crm_get_msec()` values

Reply:
Good idea, but we'll need more than that. We'll need a new function that wraps `crm_get_msec()`, since `crm_get_msec()` uses a return value of -1 (`PCMK__PARSE_INT_DEFAULT`) to indicate an error. That's a valid value for `stonith-watchdog-timeout`.

Make the new function public for the same reason as `pcmk_parse_interval_spec()`?

Reply:
Sure

Reply:
It turns out literally everything is a valid stonith-watchdog-timeout in a sense... if it's negative or can't be parsed, it's treated the same as `-1` by `pcmk__valid_sbd_timeout()`. And even that might not be applied consistently in all contexts. I'll have to do some auditing of the various uses.

On the other hand, `crm_get_msec()` treats anything negative as invalid, and it's documented as requiring a nonnegative input. So the `-1` error return value doesn't create any ambiguity. It might be helpful to have a version that can handle negative numbers, I'm not sure just yet.

A negative timeout doesn't technically make sense, but it can have special meaning. So it might be useful to differentiate between passing a negative timeout value and passing a garbage value; right now, both produce an identical negative return code.

Reply:
Ouch, I forgot a negative value is valid for stonith-watchdog-timeout. Also, for compatibility reasons we don't want to reject anything that would have been accepted by the old validator, so that means we really can't do any validation until 3.0.0 (at which point a signed integer would be appropriate).

Our current definition of timeout (and duration) in Pacemaker Explained doesn't say anything about negative values. I believe stonith-watchdog-timeout is the only option that gives a special meaning to them.

Reply:
I just went through all the cases in both the validator function and the fencer. The good news is that the validator computes the same values as the fencer and passes when the value will work in the fencer (and vice-versa). Here's a list of cases.
```
Configured value:
* negative or invalid
  * If (SBD_WATCHDOG_TIMEOUT <= 0):
    * Value: 0
    * Validation: Pass
  * Else:
    * Value: 2 * SBD_WATCHDOG_TIMEOUT
    * Validation: Pass (unless multiplication overflows or sbd not running)
* 0 or unset
  * Value: 0
  * Validation: Pass
* positive
  * Value: <configured value>
  * Validation: Pass if (value >= SBD_WATCHDOG_TIMEOUT)
                (unless multiplication overflows or sbd not running)
                Otherwise fail
    * Note: This should be "pass if (value > SBD_WATCHDOG_TIMEOUT)" (strictly greater than)
```

The core problem is that we have two default values. In that sense, stonith-watchdog-timeout is poorly specified.
* If the configured value is absent, we default to "0".
* If the configured value is invalid (i.e., a garbage string), we default to "-1".
  * "-1" is equivalent to any negative number; we say "-1" for simplicity.

Reply:
> I believe stonith-watchdog-timeout is the only option that gives a special meaning to them.

That may be. It's a legitimate use case for stonith-watchdog-timeout (to indicate "auto"), since "0" is reserved for "disabled".

Reply:
It could also make sense to drop support for negative values altogether and have a choice between the string "auto" and a timeout. That doesn't fit as neatly into the "type" field of a table or metadata, but it's arguably clearer than "negative == auto".

Reply:
> The core problem is that we have two default values. In that sense, stonith-watchdog-timeout is poorly specified.
> 
>     * If the configured value is absent, we default to "0".
> 
>     * If the configured value is invalid (i.e., a garbage string), we default to "-1".
>       
>       * "-1" is equivalent to any negative number; we say "-1" for simplicity.
> 
> 
> I think it would make more sense to default to "0" (disabled) on invalid configured values. That raises compatibility problems.

Agreed, create a task for 3.0.0

> As it stands, the default in the doc table and the options array is wrong and can't be made correct.

We'll have to live with it until then

> It could also make sense to drop support for negative values altogether and have a choice between the string "auto" and a timeout. That doesn't fit as neatly into the "type" field of a table or metadata, but it's arguably clearer than "negative == auto".

That does make more sense, but we'd need an XSL transform for existing configs. Could be part of same task. I think it would be fine to document a type like `timeout or "auto"`. However given how much else needs to be done it's probably best just to leave it as negative.

Reply:
https://projects.clusterlabs.org/T749